woocommerce/woocommerce-ios
# Order Creation: Add discard changes prompt to new order customer details

Closes: #6454 

## Description

This adds a "discard changes" prompt when swiping to dismiss the Customer Details modal on the New Order screen, to prevent accidental data loss. The prompt only appears when there are unsaved changes.

The prompt is added with a SwiftUI view modifier, which should be reusable with other SwiftUI modal views in the future.

## Changes

* Adds a SwiftUI view modifier `discardChangesPrompt(canDismiss:didDismiss:)`:
   * This modifier wraps the SwiftUI in a UIHostingController (in `DiscardChangesWrapper`), which uses `canDismiss` to determine if the modal can be dismissed with a swipe gesture and fires `didDismiss` when the modal is dismissed.
   * If `canDismiss` is false, it uses our `UIAlertController.presentDiscardChangesActionSheet` helper to display the discard changes prompt.
* Moves `hasPendingChanges()` from the `EditOrderAddressFormViewModel` (used in order editing) to `AddressFormViewModelProtocol` so it can also be used for order creation to determine if the address form has unsaved changes.
* Adds a manually created binding `canDismiss` to `CreateOrderAddressFormViewModel`. I couldn't find a better way to create this binding to use with the new view modifier; let me know if you have an alternate suggestion for this.
* Adds the view modifier to the modal in `OrderCustomerSection`, using the new binding to determine if the modal can be dismissed without the prompt.

## Testing

1. Go to the Orders tab and create a new order.
2. Select "Add Customer Details."
3. Confirm you can close the modal using the "Close" button and swiping to dismiss it before you've made any changes.
4. Add some customer details and try swiping to dismiss the modal. Confirm a "discard changes" prompt appears and works as expected.
5. Add some customer details and confirm you can tap the "Done" button to save your changes.

## Screenshots


https://user-images.githubusercontent.com/8658164/160631566-7f8e2dff-814e-4e3a-b100-a17a2fca3687.mp4



## Submitter Checklist

Update release notes:
